bitcoin/src/util/hash.rs

```diff
@@ -39,6 +39,44 @@ where
 
 /// Calculates the merkle root of an iterator of *hashes*.
 ///
+/// This fcuntion calculate the root from a merkle path when we need to derive the root from the
+/// txs list [bitcoin_merkle_root] should be used.
+///
+/// # Returns
+/// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
+/// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
+/// - `Some(merkle_root)` if length of `hashes` is greater than one.
+pub fn bitcoin_merkle_root_from_path<T, I>(mut hashes: I) -> Option<T>
+where
+    T: Hash + Encodable,
+    <T as Hash>::Engine: io::Write,
+    I: Iterator<Item = T>,
+{
+    let first = hashes.next()?;
+    let second = match hashes.next() {
+        Some(second) => second,
+        None => return Some(first),
+    };
+
+    let mut encoder = T::engine();
+    let mut root = first;
+    root.consensus_encode(&mut encoder).expect("in-memory writers don't error");
+    second.consensus_encode(&mut encoder).expect("in-memory writers don't error");
```

Comment:
You can avoid unwraps by writing `encoder.input(root.as_inner());`

Reply:
👍

Reply:
Any way to convert Inner into a byte array?

Reply:
Just dereference it? Or do you mean something else?

Reply:
In order to do that I need to restrict `Hash` to a type with a `Inner=[u8;32]` This limit the generality of this function, not seem to be an issue to me. Btw I see that the others function in the file do not do it so I want to double check it with you.

Reply:
Ah, I see now. If we want to restrict it to SHA256 at least temporarily we can do that. Or we could add `as_slice()` method to `Hash` trait. I'd normally suggest `Inner: AsRef<[u8]>` but that's not available for SHA512 in our MSRV.

Reply:
This sounds like a reason for us to at least put back `Index<range::All>`, even if we don't put back all the indexing traits.

Reply:
`Index` feels like a hack TBH. If we're breaking APIs by removing indexing we may as well move the relevant parts to a proper method(s).